### PR TITLE
[ONNX] Support optional type (#68793)

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73284

Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: garymm <garymiguel@microsoft.com>
Co-authored-by: neginraoof <neginmr@utexas.edu>

Differential Revision: [D34625646](https://our.internmc.facebook.com/intern/diff/D34625646)

### CODE REVIEW

File: torch/csrc/jit/python/python_arg_flatten.cpp
Comment (malfet) [inline torch/csrc/jit/python/python_arg_flatten.cpp:75]: This change does not seem to be specific to a PR, is it? If so, can it be moved to a separate one?

Comment (BowenBao) [reply torch/csrc/jit/python/python_arg_flatten.cpp:75]: cc original author @garymm

Comment (garymm) [reply torch/csrc/jit/python/python_arg_flatten.cpp:75]: @BowenBao do you want me to open another PR in `onnx_ms_1` that reverts the changes in this file? Or how shall I proceed?

Comment (BowenBao) [reply torch/csrc/jit/python/python_arg_flatten.cpp:75]: @garymm is it indeed not specific to this PR? Personally I do not prefer separation, since this PR already touches a lot of other files outside of ONNX as well. 

If there is a need to separate, I think the process could be you directly work on the branch of this PR, and push new commits to this branch. The separated change will be another PR submitted against master.

Comment (garymm) [reply torch/csrc/jit/python/python_arg_flatten.cpp:75]: I think it's an unrelated clean-up, but since we're getting a TorchScript code owner to review the PR anyways, I don't see any value in removing this from the PR.

Marking this thread resolved. If the TorchScript owner asks for this to be separated I can do so.

**Comment by suo** [comment]
sorry, missed the ping. @eellison do you mind looking over the JIT changes

File: aten/src/ATen/core/interned_strings.h
Comment (eellison) [inline aten/src/ATen/core/interned_strings.h:2]: Can you revert the formatting changes here in this PR or stack them ?

File: torch/csrc/jit/python/script_init.cpp
Comment (eellison) [inline torch/csrc/jit/python/script_init.cpp:466]: Why do we need the changes here ?

File: torch/csrc/jit/python/script_init.cpp
Comment (eellison) [inline torch/csrc/jit/python/script_init.cpp:495]: Why do we need changes here ?

**Review by eellison — COMMENTED** [review]
Couple questions

File: aten/src/ATen/core/interned_strings.h
Comment (BowenBao) [reply aten/src/ATen/core/interned_strings.h]: sure, reverted the formatting part.

File: torch/csrc/jit/python/script_init.cpp
Comment (BowenBao) [reply torch/csrc/jit/python/script_init.cpp:466]: To include input with OptionalType. I believe previously they are ignored.

**Comment by BowenBao** [comment]
Thanks @eellison for providing comments.

@garymm I updated the part related to reverting formatting change. I tried to answer the comments to the best of my knowledge. Please feel free to add more comments.

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34625646).

**Comment by naor2013** [comment]
@BowenBao I have a question regarding this pull request.
If I need the __is__ operator you added for "is None" use cases, can I just use the opset you added as a custom opset in PyTorch 1.11 or will it not work without the other changes you added to PyTorch?

**Comment by garymm** [comment]
@naor2013 nothing in this PR will be available in PyTorch 1.11.
You can modify an ONNX model to make use of optional types using the onnx Python API or any protocol buffer API.

**Comment by naor2013** [comment]
@garymm my question is basically, can I take the __is_ opset added in symbolic_opset15.py in this commit and use it as a custom opset or will it not work without all the other changes in this PR outside of that file?

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (thiagocrepaldi) [inline test/onnx/test_pytorch_onnx_no_runtime.py:1]: did you check whether the pipeline is calling this file?

Comment (garymm) [reply test/onnx/test_pytorch_onnx_no_runtime.py:1]: I think it is not, thanks for pointing out.  I don't want to touch this PR while it's being imported by Meta. I ran this test locally and confirms it passes, and after this is merged we can open another PR to run this automatically.

Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_no_runtime.py:1]: Tests under `test/onnx` by default runs in shard 1.

**Comment by BowenBao** [comment]
@malfet rebased again to resolve conflicts. Please import

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_no_runtime.py:96]: Is this performed by https://github.com/pytorch/pytorch/blob/a997046017aa9f97213f6bd50bc152a6445e8086/torch/testing/_internal/common_utils.py#L240 already? (Seen in instantiate_parametrized_tests's doc)

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_no_runtime.py:79]: I would include the expected behavior as part of the test method name.

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_no_runtime.py:18]: I would add your username after the todos so that readers know the right contact for this todo. (same for other todo's)

```suggestion
    # TODO(BowenBao): Move these tests to test_pytorch_onnx_onnxruntime once
```

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_no_runtime.py:96]: We'll discuss these offline, more likely update it in a follow up PR.

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (garymm) [reply test/onnx/test_pytorch_onnx_no_runtime.py:18]: blame can be used if needed. But really the comments should either be self-explanatory or contain a link that has more details. In neither case should we rely on the memory of a human.

**Comment by malfet** [comment]
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34625646).

**Comment by malfet** [comment]
This change once again regresses some of the internal model conversion tests, even though they are pinned to OpSet9 and as such should not be affected by the change, are they?

File: torch/onnx/utils.py
Comment (garymm) [reply torch/onnx/utils.py:478]: How's about:
`args = args_list if isinstance(args, list) else tuple(args_list)`
?

Thanks!

**Comment by datumbox** [comment]
@BowenBao @malfet @eellison This PR is likely to have broken TorchVision. Please see https://github.com/pytorch/vision/issues/5971 for details. Can you please revert?

**Comment by garymm** [comment]
I'll look at this today. It took us 3 months to get this PR merged so even though it's generally bad practice, I'd really like to fix forward if possible.

**Comment by datumbox** [comment]
@garymm Thanks for coming back to me. I understand it, if you want to explore fixing the issue rather than reverting the whole PR. Just please prioritise this because TorchVision's ONNX export is broken for the detection models and we are dangerous close to the release. I'm also concerned that the longer we leave this in, the harder it will be to revert it if a fix can't be completed soon.

@malfet @eellison Any thoughts/concerns on your side?

**Pull request closed without merging**
